## Data Analsyis: Anime

Welcome, coders! You've chosen to analyze the anime data set. We will be looking at the data contained in the anime.csv file, which you can check out on the left-hand side of the screen - just make your way back here when you're done.

This data set contains information on 181 different anime shows, including release data, length, genre, and rating. We're going to look at the ratings for different shows based on their genres.

First things first, we're going to have to get Python to read the file and put it into something we can work with - like a list!

Run the kernel below to do this for you.

In [35]:
# First we're going to import our helpful Python libraries

import numpy as np
import csv
from matplotlib import pyplot as plt

# We start by creating a blank list called master that we will be able to add to as the csv reader goes through the file

master = []

# Then we open the file using the csv reader and use next() to skip over the header line

with open('anime.csv', newline='') as csvfile:
    file = csv.reader(csvfile)
    next(file)
# Now we use a for loop to iterate through the file.  This reads each row as a separate item.

    for row in file:

# We use the append method on the master list to add each row as a list item as we iterate through the file.
        
        master.append(row)

Now, let's see what we're working with. In the kernel below, use the print function to print the master list.

In [12]:
# Enter code here
print(master)

[['Fullmetal Alchemist: Brotherhood', '(2009-2010)', '24 min', 'Animation, Action, Adventure', '9.1'], ['Attack on Titan', '(2013-2023)', '24 min', 'Animation, Action, Adventure', '9'], ['Hunter x Hunter', '(2011-2014)', '24 min', 'Animation, Action, Adventure', '9'], ['Bleach: Thousand-Year Blood War', '(2022- )', '24 min', 'Animation, Action, Adventure', '9'], ['Ginga eiyû densetsu', '(1988-1997)', '25 min', 'Animation, Action, Drama', '9'], ['One Piece', '(1999- )', '24 min', 'Animation, Action, Adventure', '8.9'], ['Ölüm Defteri', '(2006-2007)', '851 min', 'Animation, Crime, Drama', '8.9'], ['Cowboy Bebop', '(1998-1999)', '650 min', 'Animation, Action, Adventure', '8.9'], ['Ginga: Nagareboshi Gin', '-1986.0', '25 min', 'Animation, Action, Adventure', '8.9'], ['Vinland Saga', '(2019-2023)', '24 min', 'Animation, Action, Adventure', '8.8'], ['Dragon Ball Z', '(1989-1996)', '24 min', 'Animation, Action, Adventure', '8.8'], ['Steins;Gate', '(2011-2015)', '24 min', 'Animation, Comedy, D

Yikes, that's not very pleasant to look at, but it looks like everything is there (I hope). We don't need a lot of this information though - we're only looking for genres and ratings. So let's clean this up a little bit.

If we go back to our .csv file, we can see that 'Genre' is the fourth column and 'Rating' is the fifth. Let's try printing just the fourth and fifth indices of the master list.

In the kernel below, use the print function to display the fourth and fifth indices of master. (Remember: indexing starts at 0!)

In [16]:
# Enter code here


If you got something that starts with Bleach, you did it right! But... that's not what we're looking for.

When we added all of the data to the list, it added each row of the csv file as an item. However, each row also had 5 columns of data, so what we've got now is a bunch of lists within a list. This is ridiculist! (Sorry)

What we're looking at now is no longer a flat, one-dimensional list, but a list with two dimensions (kind of like the table we saw the data in). But fret not, Python can help us with that.

Numpy's shape function can tell us the dimensions that we're working with. Run the kernel below to determine the shape of our master list using the shape function.

In [10]:
print(np.shape(master))

(182, 5)


Hm, so this tells us that there are 182 list elements (or rows), each with 5 elements of their own (or columns). We want only the fourth and fifth item in each of the 182 elements. We can use a for loop to cycle through each element and put only the items we do want into a new list.

To reference a range of items, rather than adding them separately, I can use a colon in between two indices. For example,

my_list.append(row[1:5])

would add the elements with indices 1 through 4 (notice the 5 in the brackets means up to, but not including index 5) from the row to my_list.

The for loop is set up for you in the kernel below, but is missing the indices inside the square brackets. Fix the code to include the correct indices and then print the master list. You may move on to the next step when your output includes only the genres and ratings.

In [43]:
new_master = []
for row in master:
    new_master.append(row[:])

Well that's a bit better. However, a lot of these have multiple genres attached to them. Let's make a list for each genre that includes every rating associated with that genre. To do this, we'll need to use another for loop to iterate through the new master list, and some conditional statements.

The loop is created for you in the kernel below, however it's missing the indices. Use the clues to fill in the missing code. 
Then, use the len() function to check the lengths of each of the genre lists. You may proceed when you get the following lengths:

114 85 11 73 48

Remember: This new master list has 181 columns and only **2** rows now.

In [44]:
action = []
adventure = []
romance = []
comedy = []
drama = []

a = # Insert Genre index
b = # Insert rating index

for row in new_master:
    if 'Action' in row[a]:           
        action.append(float(row[b]))   
    if 'Adventure' in row[a]:          
        adventure.append(float(row[b]))      
    if 'Romance' in row[a]:
        romance.append(float(row[b]))
    if 'Comedy' in row[a]:
        comedy.append(float(row[b]))
    if 'Drama' in row[a]:
        drama.append(float(row[b]))
        
# Note that the float() function just converts the strings from the file to numbers so that we can work with them

<class 'SyntaxError'>: invalid syntax (<ipython-input-44-3400d87940ac>, line 7)

Awesome! Now let's do a statistical analysis. Use Numpy's quantile function and mean function to print the five-number summaries and means for each genre. Move on to the next step when your output looks like this:

Action:

Min: *number*, Q1: *number*, Median: *number*, Q3: *number*, Max: *number*, Mean: *number*

Adventure: 

Min: *number*, Q1: *number*, Median: *number*, Q3: *number*, Max: *number*, Mean: *number*

and so on...

In [ ]:
# Enter code here

Great! Now let's plot these all together and see if they match our statistical summaries. 

In the kernel below, first create a new list that has each of the genre lists as an element.

Example: new_list = [genre1, genre2, ...]

Then, create a list called labels with a string for each genre.

Example: labels = ['Genre 1', 'Genre 2', ...]

Then, use Pyplot's boxplot function to plot this new list. Include labels=labels as an input to display the genres on the plot. Use Pyplot's title, xlabel, and ylabel functions to label the graph.

In [42]:
# Enter code here

Fantastic work! You may also play around with the customization using the example in the introduction as a guide.

When you're happy with your boxplot, call me over to check your work.

Happy coding!